fili-core/src/main/java/com/yahoo/bard/webservice/table/ConcretePhysicalTable.java

```diff
@@ -20,25 +22,51 @@
 
     /**
      * Create a concrete physical table.
-     * The availability on this table is initialized to empty intervals.
      *
-     * @param name  Name of the physical table as TableName
+     * @param name  Name of the physical table as TableName, also used as fact table name
      * @param timeGrain  time grain of the table
      * @param columns  The columns for this table
      * @param logicalToPhysicalColumnNames  Mappings from logical to physical names
+     * @param metadataService  Datasource metadata service containing availability data for the table
      */
     public ConcretePhysicalTable(
             @NotNull TableName name,
             @NotNull ZonedTimeGrain timeGrain,
-            @NotNull Iterable<Column> columns,
-            @NotNull Map<String, String> logicalToPhysicalColumnNames
+            @NotNull Set<Column> columns,
+            @NotNull Map<String, String> logicalToPhysicalColumnNames,
+            @NotNull DataSourceMetadataService metadataService
+    ) {
+        this(
+                name,
+                timeGrain,
+                columns,
+                logicalToPhysicalColumnNames,
+                new ConcreteAvailability(name, columns, metadataService)
+        );
+    }
+
+    /**
+     * Create a concrete physical table, the availability on this table is built externally.
+     *
+     * @param name  Name of the physical table as TableName, also used as fact table name
+     * @param timeGrain  time grain of the table
+     * @param columns  The columns for this table
+     * @param logicalToPhysicalColumnNames  Mappings from logical to physical names
+     * @param availability  Availability that serves interval availability for columns
+     */
+    public ConcretePhysicalTable(
+            @NotNull TableName name,
+            @NotNull ZonedTimeGrain timeGrain,
+            @NotNull Set<Column> columns,
+            @NotNull Map<String, String> logicalToPhysicalColumnNames,
+            @NotNull Availability availability
```

Comment:
@archolewa There actually _is_ an additional contract that a `ConcreteAvailability` offers over an `Availability`, but it's not currently used (and perhaps it should be, and may be indicating a deeper problem). A `ConcretePhysicalTable` has this same contract extension, so let's look at what that is:

A `ConcretePhysicalTable` exposes the additional `getFactTableName`, which is used by the `DataSourceMetadataLoader` and _should_ also be used by the `TableDataSource` when we build one of those for the query. This additional method, which is there to not only indicate that there is, but also to _simplify_ retrieval of, the name of the 1 and only 1 backing data source. Non-concrete physical tables cannot make this assertion, and it's an important one to have (1 vs. many).

Now, the possibly _smelly_ part of the current implementation is that `ConcretePhysicalTable` uses _it's_ name as the name of it's backing data source, rather than getting it from it's contained `ConcreteAvailability`. This is probably incorrect (ie. it _should_ call down to the concrete availability to get it), and the following use case illustrates it:

Consider that we have a single Druid data source that contains both hourly and daily data, where it's hourly for the 1st 30 days, and then switches to daily from then on. If we wanted to be able to have Fili's table selection logic see that table as both an hourly table _and_ a daily table, we need to have 2 different physical tables. Here's where the problem arises: Only _one_ of those two tables (based on the current implementation) can have it's name match the name of Druid's data source. In order for both tables to have the same underlying data source, the only way to get there would be to have it sourced from the same `ConcreteAvailability`, which would need to have the equivalent singleton method.

So, given the above, not only should we _keep_ `ConcreteAvailability`, but we should also fix the abstraction that `ConcretePhysicalTable` is giving us, by making it source it's `factTableName` from it's underlying `ConcreteAvailability`.

<sub>P.S. Sorry for the short saga, I'm in a storytelling mood I guess 😉 </sub>